### PR TITLE
Support Predicate Pushdown for Parquet Lists (#2108)

### LINKED ISSUE #2108: Support skip_rep_levels in ColumnLevelDecoder
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
https://github.com/apache/arrow-rs/blob/c58554453bec8f4a78d503e79576b85108b210b5/parquet/src/column/reader/decoder.rs#L84-L90
**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #2108

# Rationale for this change
 

This PR adds support for skipping repetition levels by adding a buffer to ColumnLevelDecoderImpl. This buffer is also used to speed up skipping definition levels, as it allows for vectorised decoding.

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

**Comment by tustvold** [comment]
Will update with rep_levels support

**Comment by tustvold** [comment]
https://arrow.apache.org/blog/2022/10/08/arrow-parquet-encoding-part-2/ may be helpful here

**Comment by Ted-Jiang** [comment]
I will review this carefully tomorrow morning!

File: parquet/src/column/reader/decoder.rs [hunk 1]
```diff
@@ -305,12 +337,25 @@ impl ColumnLevelDecoder for ColumnLevelDecoderImpl {
         }
     }
 
-    fn read(&mut self, out: &mut Self::Slice, range: Range<usize>) -> Result<usize> {
+    fn read(&mut self, out: &mut Self::Slice, mut range: Range<usize>) -> Result<usize> {
+        let read_from_buffer = match self.buffer.is_empty() {
```
Comment (Ted-Jiang) [inline parquet/src/column/reader/decoder.rs:341]: 👍

File: parquet/src/column/reader/decoder.rs [hunk 2]
```diff
@@ -323,41 +368,153 @@ impl DefinitionLevelDecoder for ColumnLevelDecoderImpl {
...
 impl RepetitionLevelDecoder for ColumnLevelDecoderImpl {
-    fn skip_rep_levels(&mut self, _num_records: usize) -> Result<(usize, usize)> {
-        Err(nyi_err!("https://github.com/apache/arrow-rs/issues/1792"))
+    fn skip_rep_levels(&mut self, num_records: usize) -> Result<(usize, usize)> {
+        let mut level_skip = 0;
+        let mut record_skip = 0;
+
+        loop {
+            if self.buffer.is_empty() {
+                // Read SKIP_BUFFER_SIZE as we don't know how many to read
```
Comment (Ted-Jiang) [inline parquet/src/column/reader/decoder.rs:404]: 👍

File: parquet/src/column/reader/decoder.rs [hunk 3]
```diff
@@ -323,41 +368,153 @@ impl DefinitionLevelDecoder for ColumnLevelDecoderImpl {
...
+            while to_skip < self.buffer.len() && record_skip != num_records {
+                if self.buffer[to_skip] == 0 {
+                    record_skip += 1;
+                }
+                to_skip += 1;
+            }
+
+            // Find end of record
+            while to_skip < self.buffer.len() && self.buffer[to_skip] != 0 {
+                to_skip += 1;
```
Comment (Ted-Jiang) [inline parquet/src/column/reader/decoder.rs:422]: Nice check !

File: parquet/src/column/reader/decoder.rs [hunk 4]
```diff
@@ -323,41 +368,153 @@ impl DefinitionLevelDecoder for ColumnLevelDecoderImpl {
...
+            }
+
+            level_skip += to_skip;
+            if to_skip >= self.buffer.len() {
```
Comment (Ted-Jiang) [inline parquet/src/column/reader/decoder.rs:426]: I think here only need `==`

File: parquet/src/column/reader/decoder.rs [hunk 5]
```diff
@@ -323,41 +368,153 @@ impl DefinitionLevelDecoder for ColumnLevelDecoderImpl {
...
+        F: Fn(&mut ColumnLevelDecoderImpl, &mut usize, usize),
+    {
+        let mut rng = thread_rng();
+        let mut decoder = ColumnLevelDecoderImpl::new(5);
+        decoder.set_data(Encoding::RLE, data);
+
+        let mut read = 0;
+        let mut decoded = vec![];
+        let mut expected = vec![];
+        while read < encoded.len() {
```
Comment (Ted-Jiang) [inline parquet/src/column/reader/decoder.rs:457]: Nice fuzz tests !  👍

**Review by Ted-Jiang — APPROVED** [review]
Nice improvement! ❤️ @tustvold

File: parquet/src/column/reader/decoder.rs [hunk 6]
```diff
@@ -264,9 +264,13 @@ impl<T: DataType> ColumnValueDecoder for ColumnValueDecoderImpl<T> {
     }
 }
 
+const SKIP_BUFFER_SIZE: usize = 1024;
```
Comment (Ted-Jiang) [inline parquet/src/column/reader/decoder.rs:267]: @tustvold choose `SKIP_BUFFER_SIZE: usize = 1024`,  1024* i16 just 2Kb,  which supported by AVX_256, so this is the best choice 🤔 (i am not familiar with this, could you give me some inforamtion😂)

Comment (tustvold) [reply parquet/src/column/reader/decoder.rs:267]: It's a somewhat arbitrary number, it tries to strike a balance between small enough to not be overly wasteful but large enough that we don't end up repeatedly reading too few values

File: parquet/src/column/reader/decoder.rs (hunk 4 above)
Comment (tustvold) [reply parquet/src/column/reader/decoder.rs:426]: Yeah, the hope is that >= helps LLVM elide bound checks that follow. Not sure it makes any difference, just habit 😅

**Pull request merged**